### PR TITLE
Add upgrade actions to the helm release list

### PR DESCRIPTION
**Pull request by debsmita1**
**Fixes:**
https://issues.redhat.com/browse/ODC-2525

**Analysis/Root cause:**
Add provision to upgrade the helm releases

**Solution Description:**
- This PR adds Helm Release Resource Icon and Chart Version column in the Helm Releases list page
- Adds a kebab menu in the Helm Release Rows having a menu action to Upgrade the Helm Release, Upgrade action is added in the Actions dropdown of the Helm Release Details page, in the Topology context menu & side panel when helm group is selected
- Upgrade Helm Release form has the Release Name which cannot be altered by the user, has a dropdown which lists all the Chart Versions available for the Helm Release and the Yaml editor to edit values.yaml of the Helm Release. Chart Version dropdown is disabled for helm releases installed via cli
- on form submission, PUT request is sent to the `/api/helm/release` with payload - Helm Release Name, namespace, and values dictionary.
- added test case for the newly added util.

**Depends on the backend api** - **PUT /api/helm/release** (PR: https://github.com/openshift/console/pull/4580)

**GIF/Screenshots:**
helm-release-upgrade

Chart version is disabled for Helm Releases that are installed via CLI
Screenshot from 2020-03-26 20-12-58

**Unit Tests:**
HIUF

helm-utils

Note: In the above GIF & screenshot the IBM charts index.yaml has been used (https://raw.githubusercontent.com/IBM/charts/master/repo/community/index.yaml) .

### CODE REVIEW

**Comment by rohitkrai03** [comment]
@debsmita1 Please remove the commits from the backend PR. You don't need that in your UI PR. You just need to build your bridge from that PR in order to test the UI changes on top of the backend.

File: frontend/packages/dev-console/src/plugin.tsx [hunk 1]
```diff
@@ -627,6 +627,19 @@ const plugin: Plugin<ConsumedExtensions> = [
         ).default,
     },
   },
+  {
+    type: 'Page/Route',
+    properties: {
+      exact: true,
+      path: [`/helm-releases/ns/:ns/:releaseName/upgrade-helm-release`],
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/plugin.tsx]:
```suggestion
      path: [`/helm-releases/ns/:ns/:releaseName/upgrade`],
```

File: frontend/packages/dev-console/src/components/helm/helm-types.ts [hunk 2]
```diff
@@ -39,6 +39,20 @@ export interface HelmChart {
   schema?: string;
 }
 
+export interface HelmChartData {
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/helm-types.ts]: This is exactly same as `HelmChart.metadata` from above interface.

File: frontend/packages/dev-console/src/components/helm/HelmReleaseRow.tsx [hunk 3]
```diff
@@ -34,9 +43,15 @@ const HelmReleaseRow: React.FC<HelmReleaseRowProps> = ({ obj, index, key, style
...
       </TableData>
       <TableData className={tableColumnClasses.chartName}>{obj.chart.metadata.name}</TableData>
+      <TableData className={tableColumnClasses.chartVersion}>
+        {obj.chart.metadata.version}
+      </TableData>
       <TableData className={tableColumnClasses.appVersion}>
         {obj.chart.metadata.appVersion}
       </TableData>
+      <TableData className={tableColumnClasses.kebab}>
+        <Kebab options={menuActions} key={`${obj.name}-${obj.version}`} />
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmReleaseRow.tsx]: This PR adds the upgrade action to just the details page. But we also have helm release details page and topology where we show helm release and actions need to be added there as well.

Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseRow.tsx]: I know the story AC doesn't explicitly mentions this but we did discuss this during the grooming of the stories and we updated the AC of `delete` story. Maybe we missed updating the AC for this one.

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+import { ALL_NAMESPACES_KEY } from '@console/shared';
+import { coFetchJSON, coFetch } from '@console/internal/co-fetch';
+import { history, PageHeading, LoadingBox } from '@console/internal/components/utils';
+import NamespacedPage, { NamespacedPageVariants } from '../NamespacedPage';
+import { HELMRELEASES_ALL_NS_PAGE_URI } from './HelmReleaseDetails';
+import HelmReleaseUpgradeForm from './HelmReleaseUpgradeForm';
+import { HelmRelease, HelmChartData } from './helm-types';
+import { getChartURL } from './helm-utils';
+
+const handleNamespaceChange = (newNamespace: string): void => {
```
[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: The UXD doesn't have the namespace selector disabled

[the remark this answers is not in the document]
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: But you have it disabled in your code. You are setting the variant as light there.

[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: If I am not wrong, the `NamespacedPageVariants` takes care of the background of the area that is below the Namespace Bar.

[the remark this answers is not in the document]
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: Ah, my mistake. You need to pass `disabled` prop to disable it. But I am not sure what is the reasoning for UXD to enabled the change of namesapce in upgrade page when it is disabled everywhere else in other flows. Also, i am not sure how the changing the namespace of release would work in Helm. Would it delete the release in current namespace and create a new release in another? IMO it should be disabled.

[the remark this answers is not in the document]
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: @serenamarie125 @parvathyvr Can you guys weigh in on this as to why the namespace bar is enabled in helm release upgrade form?

[the remark this answers is not in the document]
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: @debsmita1 What is the value of having the namespace bar enabled when all it does is redirect the user to the helm release list page?

**Comment by rohitkrai03** [comment]
The dropdown doesn't handle the case where there are no available chart versions - 
Screenshot from 2020-03-23 16-48-10

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: In the Build Configs/Builds/Deployments pages, if you choose to edit a resource it will take you to the YAML editor. 
The Namespace bar is not disabled, and if you go ahead and change the namespace it will take you the resources list page. I replicated this behavior in the Upgrade Helm Release Form as well.

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    if (chartValuesYAML) {
+      try {
+        valuesObj = safeLoad(chartValuesYAML);
+      } catch (err) {
+        actions.setStatus({ submitError: `Invalid YAML - ${err}` });
+        return;
+      }
+    }
+    const chartURL: string = getChartURL(helmChartData, chartVersion);
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: The upgrade will fail for any chart that was installed via CLI or some other catalog since you assume that you'll find the `chartURL` in one of the charts available in the catalog.

**Comment by parvathyvr** [comment]
@debsmita1 I agree with @rohitkrai03 about disabling the project drop down in the secondary navigation.It should be disabled when the user is in the 'Upgrade Helm Release' form like in the other +Add flow forms.It was a mistake in the mockups which is corrected now.

**Comment by debsmita1** [comment]
> The dropdown doesn't handle the case where there are no available chart versions -
> Screenshot from 2020-03-23 16-48-10

@rohitkrai03 Did you install that particular chart from the helm cli ? The Chart version dropdown will be populated only if that particular chart is present in the index.yaml that is being used to fetch the versions. In that case dropdown will be empty

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (serenamarie125) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgrade.tsx]: > Ah, my mistake. You need to pass disabled prop to disable it. But I am not sure what is the reasoning for UXD to enable the change of namesapce in upgrade page when it is disabled everywhere else in other flows. Also, i am not sure how the changing the namespace of release would work in Helm. Would it delete the release in current namespace and create a new release in another? IMO it should be disabled.

@debsmita1 @rohitkrai03 we missed that!  It's too bad this didn't come up in the review :). The project selector bar should remain disabled with this flow.

File: frontend/packages/dev-console/src/components/helm/helm-types.ts [hunk 6]
```diff
@@ -19,26 +19,29 @@ export interface HelmRelease {
...
-    appVersion: string;
-    keywords?: string[];
-    home?: string;
-    icon?: string;
-    sources?: string[];
-    maintainers?: object[];
-    dependencies?: object[];
-    type?: string;
-  };
+  metadata: HelmChartData;
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/helm-types.ts]:
```suggestion
  metadata: HelmChartMetaData;
```

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
+import * as React from 'react';
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: `HelmActionForm` might be misleading here since we have a lot more Helm actions than install and upgrade. Probably better to name it `HelmInstallUpgradeForm`. WDYT?

File: frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx [hunk 8]
```diff
@@ -1,124 +1,55 @@
...
-      </PageBody>
-    </NamespacedPage>
+    <HelmActionPage
+      namespace={preselectedNamespace}
+      title={'Install Helm Chart'}
+      pageDescription={'The helm chart will be installed using the YAML shown in the editor below.'}
+      chartName={chartName}
+      chartURL={chartURL}
+      helmAction={HelmActionType.Install}
+      historyURL={`/topology/ns/${preselectedNamespace}`}
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx]:
```suggestion
      redirect={`/topology/ns/${preselectedNamespace}`}
```

File: frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx [hunk 9]
```diff
@@ -1,124 +1,55 @@
...
+      chartValues={!_.isEmpty(chartData.values) && chartData.values}
+      helmActionRequest={helmInstallRequest}
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx]:
```suggestion
      onSubmit={helmInstallRequest}
```

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx [hunk 10]
```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: Rename this one for consistency. `HelmUpgradePage`.

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  const [helmChartData, setHelmChartData] = React.useState<HelmChartMetaData[]>([]);
+  const [chartValuesLoaded, setChartValuesLoaded] = React.useState(false);
+  const [chartVersions, setChartVersions] = React.useState({});
+  const [activeChartVersion, setActiveChartVersion] = React.useState<string>('');
+
+  React.useEffect(() => {
+    const getChartDetails = () => {
+      coFetchJSON(`/api/helm/release?ns=${namespace}&release_name=${releaseName}`)
+        .then((res) => {
+          setActiveChartVersion(res.chart.metadata.version);
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: Possible data leak when the component is unmounted. Follow the older implementation.

File: frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx [hunk 12]
```diff
@@ -1,124 +1,55 @@
...
-      }
-      if (ignore) return;
-
-      setChartData(res);
-      setChartDataLoaded(true);
-      !_.isEmpty(res.values) && setChartHasValues(true);
+    const fetchHelmReleases = () => {
+      coFetchJSON(`/api/helm/chart?url=${chartURL}`)
+        .then((res) => {
+          setChartData(res);
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx]: Possible data leak here. Revert back to the older implementation.

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+          json = safeLoad(yaml);
+          const versions = _.reduce(
+            _.get(json, ['entries', chartName]),
+            (obj, chart) => {
+              obj[chart.version] = chart.version;
+              return obj;
+            },
+            {},
+          );
+          setHelmChartData(_.get(json, ['entries', chartName]));
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: Same as above. Avoid data leaks.

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        })
+        .catch((err) => {
+          // eslint-disable-next-line no-console
+          console.error(err);
+        });
+    };
+    getChartVersions();
+  }, [chartName]);
+
+  const onVersionChange = (value: string) => {
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: Why do you need this logic here in this component. This could easily go into the dropdown component for chart version.

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+      releaseName={releaseName}
+      title={'Upgrade Helm Release'}
+      helmAction={HelmActionType.Upgrade}
+      chartName={chartName}
+      chartVersions={chartVersions}
+      activeChartVersion={activeChartVersion}
+      historyURL={`/helm-releases/ns/${namespace}`}
+      chartValues={chartValues}
+      onChartVersionChange={onVersionChange}
+      helmActionRequest={helmUpgradeRequest}
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]:
```suggestion
      onSubmit={helmUpgradeRequest}
```

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+      chartURL={getChartURL(helmChartData, activeChartVersion)}
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: Why is this needed to be passed from here? I though the backend is now supporting request without `chartURL`.

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+import { coFetchJSON } from '@console/internal/co-fetch';
+import { InputField, FormFooter, FlexForm, YAMLEditorField, DropdownField } from '@console/shared';
+import FormSection from '../import/section/FormSection';
+import { HelmActionType } from './helm-types';
+
+interface HelmActionFormProps {
+  chartHasValues: boolean;
+  chartVersions?: {};
+  activeChartVersion?: string;
+  onChartVersionChange?: (value: string) => string;
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: Don't think you this this a prop.

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+              name="helmReleaseName"
+              label="Release Name"
+              helpText="A unique name for the Helm Chart release."
+              required
+              isDisabled={!_.isEmpty(activeChartVersion)}
+            />
+          </GridItem>
+          {activeChartVersion && (
+            <GridItem span={6}>
+              <DropdownField
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: You should create a separate component for this.

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+                name="chartVersion"
+                label="Chart Version"
+                items={chartVersions}
+                fullWidth
+                helpText={'Select the version to upgrade to.'}
+                required
+                disabled={_.isEmpty(chartVersions)}
+                title={activeChartVersion}
+                onChange={(value) => {
+                  const url = onChartVersionChange(value);
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: Why not just use `getChartURL` util function here?

File: frontend/packages/dev-console/src/components/helm/HelmActionPage.tsx [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+import { PageHeading, history } from '@console/internal/components/utils';
+import { PageBody } from '@console/shared';
+import NamespacedPage, { NamespacedPageVariants } from '../NamespacedPage';
+import { nameValidationSchema } from '../import/validation-schema';
+import HelmActionForm from './HelmActionForm';
+import { HelmActionType } from './helm-types';
+
+export type HelmActionPageProps = {
+  namespace: string;
+  title: string;
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmActionPage.tsx]: No need of this prop. You can use `helmAction` prop to create the title.

File: frontend/packages/dev-console/src/components/helm/HelmActionPage.tsx [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+  pageDescription?: string;
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmActionPage.tsx]: No need for this prop as well.

File: frontend/packages/dev-console/src/components/helm/HelmActionPage.tsx [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+  helmActionRequest: (object) => Promise<any>;
+};
+
+export type HelmActionFormData = {
+  helmReleaseName: string;
+  chartValuesYAML: string;
+  chartVersion?: string;
+};
+
+export const getValidationSchema = (helmAction: HelmActionType) => {
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmActionPage.tsx]: Move this to utils file or `helm-validation-utils.ts`

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx (hunk 7 above)
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: Ignore this. It would create another confusion. We already have `HelmInstallPage` and `HelmUpgradePage`.

Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: Removed `HelmInstallPag` & `HelmReleaseUpgradePage` and have created one common `HelmInstallUpgradePage`

File: frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx (hunk 8 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx]: removed this prop

File: frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx (hunk 9 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmInstallPage.tsx]: removed it

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx (hunk 10 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: removed this component

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx (hunk 14 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: I have kept it in the HelmInstallUpgradePage, as there I am fetching the entries of a helm chart, so every time user updates the version from the dropdown, to be able to fetch the values.yaml of that selected version I must fetch the url for that version. please do let me know if you still have a different opinion

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx (hunk 15 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: removed  it

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx (hunk 16 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: removed it

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx (hunk 17 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: same as https://github.com/openshift/console/pull/4785#discussion_r399575853

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx (hunk 19 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: same as https://github.com/openshift/console/pull/4785#discussion_r399575853

**Comment by serenamarie125** [comment]
In a future PR ( UX would need to help here ) we should consider adding help if the Chart Version is disabled.  It seems like there may be a couple of cases here : (1) there is no upgradable version and (2) if it was installed via CLI we may not be able to support it.  @parvathyvr as part of the Final Design story, let's dig into this a bit more!  I'll add a comment in the DTUX story.  Posting this here FYI @debsmita1

**Comment by debsmita1** [comment]
> @debsmita1 this looks good! Could you verify a few things? I see that you added this in the kebab menu in the list view. Can you confirm that this is also available from (1) the Actions menu in the Helm Release Details View and (2) the Actions menu in the side panel in Topology when a Helm Release is selected?
> 
> Thanks!

@serenamarie125 
Upgrade action is added in the Actions dropdown of the Helm Release Details page:
Screenshot from 2020-03-30 10-35-12

Upgrade Action is added in the Topology context menu & side panel when helm group is selected
Screenshot from 2020-03-30 10-35-02
Screenshot from 2020-03-30 10-34-56

File: frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx (hunk 14 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmReleaseUpgradePage.tsx]: Also, in handleSubmit the helm chart entries is being used to fetch the chart url in case user wants to update the chart version

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx (hunk 18 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: Created `HelmChartVersionDropdown` component

File: frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx (hunk 19 above)
Comment (debsmita1) [reply frontend/packages/dev-console/src/components/helm/HelmActionForm.tsx]: Made the requested change

File: frontend/packages/dev-console/src/components/helm/HelmInstallUpgradePage.tsx [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+      name: helmReleaseName,
+      ...(helmAction === HelmActionType.Install
+        ? // eslint-disable-next-line @typescript-eslint/camelcase
+          { chart_url: chartURL }
+        : // eslint-disable-next-line @typescript-eslint/camelcase
+          { chart_url: newChartURL }),
+      ...(valuesObj ? { values: valuesObj } : {}),
+    };
+
+    coFetchSendJSON(
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmInstallUpgradePage.tsx]: I think you can use `coFetchJSON` and send the method similar to this as well. That way you won't need to have change in console code. This function just adds some extra headers, I doubt if those headers are absolutely necessary.

File: frontend/packages/dev-console/src/components/helm/HelmInstallUpgradePage.tsx [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+        return;
+      }
+    }
+    let newChartURL = '';
+    newChartURL = chartVersion && getChartURL(helmChartEntries, chartVersion);
+
+    const payload = {
+      namespace,
+      name: helmReleaseName,
+      ...(helmAction === HelmActionType.Install
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmInstallUpgradePage.tsx]: Why don't you also save `chartURL` in formik state? Add it in the initial values and update it from the `ChartVersionDropdown` when user selects a new version. That way you don't have to have any logic here and just use `chartURL` from the formik state here.

File: frontend/packages/dev-console/src/components/helm/HelmChartVersionDropdown.tsx [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+    setHelmChartVersions(chartVersions);
+  }, [chartEntries]);
+
+  const onChartVersionChange = (value: string) => {
+    setChartVersion(value);
+    setFieldValue('chartVersion', value);
+    const url = getChartURL(chartEntries, value);
+    coFetchJSON(`/api/helm/chart?url=${url}`)
+      .then((res) => {
+        setFieldValue('chartValuesYAML', safeDump(res.values));
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmChartVersionDropdown.tsx]: Maybe add a safe check here before setting the value similar to what is there while setting initial values in `HelmInstallUpgradePage`.

File: frontend/packages/dev-console/src/components/helm/HelmInstallUpgradePage.tsx (hunk 23 above)
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/helm/HelmInstallUpgradePage.tsx]: You can get the `cofetchJSON.post` and `coFetchJSON.put` by doing `_.get(coFetchJSON, 'post')`. This way you can avoid using `coFetchSendJSON` directly in your code and still have the logic to use correct method based on the helm action.

File: frontend/packages/dev-console/src/components/helm/HelmInstallUpgradePage.tsx [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+      redirectURL: `/topology/ns/${namespace}`,
+    },
+    Upgrade: {
+      title: 'Upgrade Helm Release',
+      helmReleaseApi: `/api/helm/release?ns=${namespace}&release_name=${releaseName}`,
+      fetch: coFetchJSON.put,
+      redirectURL: `/helm-releases/ns/${namespace}`,
+    },
+  };
+
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/helm/HelmInstallUpgradePage.tsx]: Why not just add `const config = HelmActionConfig[helmAction];` and then use `config.something` everywhere without having to do the boolean check again and again?
